PR: https://github.com/odoo/odoo/pull/259842

From: d069ce59e28fda2bc25fb89fd01ee7b988a30fa0
From: Nisarg (nipl)
Date: 2026-06-15 17:24:30

Total Changes: 142

[IMP] sale, account, *: support description-first flow on sale orders and invoices

This commit introduces support for productless Sale Order Lines (SOLs),
allowing users to create quotation/invoice lines directly from descriptions
without requiring a product.

Functional changes

* For small business workflows, creating products for every line can be
  unnecessary overhead. SOLs can now be created without a product while
  preserving standard sale and invoicing flows.
* A new setting allows companies to keep the previous product-first behavior
  by making products mandatory on SOLs.
* When products are mandatory, UoM selection remains restricted to the
  product's compatible UoMs.
* For productless SOLs, users can select any UoM.
* All standard SOLs (`display_type=False`) now default to `Units` as UoM.
  Once a product is selected, the UoM is automatically adjusted according
  to the product configuration.
* New productless SOLs now receive the company default sales taxes by
  default, while still allowing manual tax customization.
* Automatic Invoicing setting is now converted to company specific setting
  instead of config param.
* When applying quotation templates on sale orders now created SOLs retain
  product name + (template) description, for consistancy on both the places
* Productless SOLs are now fully supported in section templates:
  taxes, unit price, discount, and unit cost are preserved when saving
  and reapplying templates (currency conversions are also handled).
* Loyalty programs having no product restriction takes productless SOLs
  in cosideration when calculating rewards.
* In sales reports productless SOLs are displayed as their name instead
  of product name (with selected UOM).
* For Commissions and achievements rules which have no product restrictions
  directly aggregates productless SOLs.
* To keep consistant portal flow, eCommerce links to products are now only
  shown for eCommerce orders. (because productless lines won't have the link
  and it shouldn't be user's concern whether some product is there in our DB
  or not).
* When creating the same product as one already added through a quotation
  template, we previously reused the template's description. This is not
  very useful, as if the user wants to sell the product under the same
  description, they would typically just increase the quantity instead.

UI & UX

* Description field of SO and AM  now acts as the main product + description
  textarea and supports direct product search from the description itself,
  without requiring interaction with the product field.
* When a product is readonly, editing the description no longer opens the
  product autocomplete, ensuring users can safely edit labels without
  modifying the product.
* Sale configurators and combo flows continue to behave as before on new field.
* Any configuration or combo flow using the description field column now
  automatically uses the new searchable description field widget.
* Edit Configuration button is now only visible on hover, for better UX.
* Also stacked description is now always visible, the button for toggle its
  visibility is completely removed.
* When mandatory product feature is active, users won't be able to search on
  description field, for mandatory product mode, users must switch to old
  Product field.
* Edit button UX is changed bit to only show on hover (and when line is not
  edit mode).
* Removed `X {quantity}` from combo product line's display_name since we
  already display quantity field on combo line and it's pointless code to
  maintain. (eg. Office Combo X 3 -> Office Combo)
* Description now completely reset through onchange when user removes
  product from exisiting line.
* For managing productless lines in quotation/section templates we introduced
  optional hidden columns unit price, discount and taxes on template lines,
  which are invisible for lines having product, and should be selectable when
  atleast one productless line is added in template.

Technical changes

* `default_invoice_policy` was converted to `sale_invoice_policy`, a
  company-dependent field.
* For SOL invoice policy:
  * product SOLs use `product_id.invoice_policy`
  * productless SOLs fallback to `company.sale_invoice_policy`
* For searching invoice policy of any SOL (with or without product)
  a new computed search field `sale.order.line.invoice_policy` is introduced.
* Extracted combos and configurators logic in `saleProductMixin` which contains
  methods which should be called after `product.template` or `product.product`
  updates rest of the things are handled in mixin itself.
* Cleaned Up and improved some `sale` tour utils such as
  * reduced indeterministic behaviours by adding a `o_has_product` flag on
    description field.
  * Added logic to check multiline description of some SOL.
  * Refined some selectors of tours to gain better certainity on element
    we want to select.
* We also noticed that `product_id` was not removed in onchange when removing
  product from `product_template_id` field, which caused some issues like
  old name still showing there or having `product_id` w/o `product_template_id`.
  Added onchange to make both the fields consistant.

task-6109825

Part-of: odoo/odoo#259842
Related: odoo/enterprise#114608
Related: odoo/upgrade#9981
Signed-off-by: Victor Feyens (vfe) <vfe@odoo.com>
Co-authored-by: Nisarg Patel (nipl) <nipl@odoo.com>
Co-authored-by: Fabien Pinckaers (fp) <fp@openerp.com>
Co-authored-by: John Laterre (jol) <jol@odoo.com>
Co-authored-by: Julien Carion (juca) <juca@odoo.com>

================================= pseudo patch: =================================

--- a/addons/sale/models/payment_transaction.py
+++ b/addons/sale/models/payment_transaction.py
@@ -99,7 +99,7 @@ class PaymentTransaction(models.Model):
                 confirmed_orders = done_tx._check_amount_and_confirm_order()
                 (done_tx.sale_order_ids - confirmed_orders)._send_payment_succeeded_for_order_mail()
 
-            auto_invoice = self.env["ir.config_parameter"].sudo().get_bool("sale.automatic_invoice")
+            auto_invoice = done_tx.company_id.sale_automatic_invoice
             if auto_invoice:
                 # Invoice the sales orders of confirmed transactions instead of only confirmed
                 # orders to create the invoice even if only a partial payment was made.
@@ -179,9 +179,6 @@ class PaymentTransaction(models.Model):
 
     def _cron_send_invoice(self):
         """Send invoices that where not ready to be send directly after posting."""
-        if not self.env["ir.config_parameter"].sudo().get_bool("sale.automatic_invoice"):
-            return
-
         # No need to retrieve old transactions
         retry_limit_date = datetime.now() - relativedelta.relativedelta(days=2)
         # Retrieve all transactions matching the criteria for post-processing
@@ -198,6 +195,7 @@ class PaymentTransaction(models.Model):
             ),
             ("sale_order_ids.state", "=", "sale"),
             ("last_state_change", ">=", retry_limit_date),
+            ("company_id.sale_automatic_invoice", "=", True),
         ])._send_invoice()
 
     def _invoice_sale_orders(self):

--- a/addons/sale/models/product_template.py
+++ b/addons/sale/models/product_template.py
@@ -5,6 +5,8 @@ from collections import defaultdict
 from odoo import api, fields, models
 from odoo.exceptions import ValidationError
 
+from odoo.addons.sale.models.res_company import SALE_INVOICE_POLICY
+
 
 class ProductTemplate(models.Model):
     _inherit = "product.template"
@@ -39,7 +41,7 @@ class ProductTemplate(models.Model):
     )
     sales_count = fields.Float(string="Sold", compute="_compute_sales_count", digits="Product Unit")
     invoice_policy = fields.Selection(
-        selection=[("order", "Ordered quantities"), ("delivery", "Delivered quantities")],
+        selection=SALE_INVOICE_POLICY,
         string="Invoicing Policy",
         compute="_compute_invoice_policy",
         precompute=True,
@@ -206,7 +208,11 @@ class ProductTemplate(models.Model):
 
     @api.depends("type")
     def _compute_invoice_policy(self):
-        self.filtered(lambda t: t.type == "consu" or not t.invoice_policy).invoice_policy = "order"
+        for template in self:
+            if not template.invoice_policy:
+                template.invoice_policy = self.env.company.sale_invoice_policy
+            elif template.type == "consu":
+                template.invoice_policy = "order"
 
     def _get_backend_root_menu_ids(self):
         return super()._get_backend_root_menu_ids() + [self.env.ref("sale.sale_menu_root").id]

--- a/addons/sale/models/res_company.py
+++ b/addons/sale/models/res_company.py
@@ -3,6 +3,8 @@
 from odoo import api, fields, models
 from odoo.exceptions import ValidationError
 
+SALE_INVOICE_POLICY = [("order", "Ordered quantities"), ("delivery", "Delivered quantities")]
+
 
 class ResCompany(models.Model):
     _inherit = "res.company"
@@ -56,6 +58,21 @@ class ResCompany(models.Model):
         tracking=True,
     )
 
+    sale_invoice_policy = fields.Selection(
+        selection=SALE_INVOICE_POLICY, string="Invoicing Policy", default="order", required=True
+    )
+
+    sale_automatic_invoice = fields.Boolean(
+        string="Automatic Invoicing",
+        help="The invoice is generated automatically and available in the customer portal when the "
+        "transaction is confirmed by the payment provider.\nThe invoice is marked as paid and "
+        "the payment is registered in the payment journal defined in the configuration of the "
+        "payment provider.\nThis mode is advised if you issue the final invoice at the order "
+        "and not after the delivery.",
+    )
+
+    # === CONSTRAINT METHODS === #
+
     @api.constrains("prepayment_percent")
     def _check_prepayment_percent(self):
         for company in self:
@@ -63,3 +80,24 @@ class ResCompany(models.Model):
                 raise ValidationError(
                     company.env._("Prepayment percentage must be a valid percentage.")
                 )
+
+    # === CRUD METHODS ===#
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        result = super().create(vals_list)
+        self._sync_automatic_invoice_cron()
+        return result
+
+    def write(self, vals):
+        result = super().write(vals)
+        if "sale_automatic_invoice" in vals:
+            self._sync_automatic_invoice_cron()
+        return result
+
+    def _sync_automatic_invoice_cron(self):
+        automatic_invoice_cron = self.env.ref("sale.send_invoice_cron", raise_if_not_found=False)
+        if not automatic_invoice_cron:
+            return
+        active_companies = self.search_count([("sale_automatic_invoice", "=", True)], limit=1)
+        automatic_invoice_cron.sudo().active = bool(active_companies)

--- a/addons/sale/models/sale_order.py
+++ b/addons/sale/models/sale_order.py
@@ -1037,7 +1037,7 @@ class SaleOrder(models.Model):
                 continue
             dates_list = order.order_line.filtered(
                 lambda line: (
-                    line.product_id.type == "consu"
+                    (not line.product_id or line.product_id.type == "consu")
                     and not line.display_type
                     and not line._is_delivery()
                 )
@@ -1572,14 +1572,6 @@ class SaleOrder(models.Model):
         self.ensure_one()
         if self.state not in {"draft", "sent"}:
             return self.env._("Some orders are not in a state requiring confirmation.")
-        if any(
-            not line.display_type and not line.is_downpayment and not line.product_id
-            for line in self.order_line
-        ):
-            return self.env._(
-                "Some order lines are missing a product, you need to correct them before going"
-                " further."
-            )
 
         return False
 

--- a/addons/sale/models/sale_order_line.py
+++ b/addons/sale/models/sale_order_line.py
@@ -18,7 +18,7 @@ class SaleOrderLine(models.Model):
     _check_company_auto = True
 
     _accountable_required_fields = models.Constraint(
-        "CHECK(display_type IS NOT NULL OR is_downpayment OR (product_id IS NOT NULL AND product_uom_id IS NOT NULL))",  # noqa: E501
+        "CHECK(display_type IS NOT NULL OR is_downpayment OR product_uom_id IS NOT NULL)",
         "Missing required fields on accountable sale order line.",
     )
     _non_accountable_null_fields = models.Constraint(
@@ -160,8 +160,8 @@ class SaleOrderLine(models.Model):
     product_uom_id = fields.Many2one(
         comodel_name="uom.uom",
         string="Unit",
+        domain="[('id', 'in', allowed_uom_ids)] if allowed_uom_ids or mandatory_product else []",
         compute="_compute_product_uom_id",
-        domain='[("id", "in", allowed_uom_ids)]',
         store=True,
         readonly=False,
         precompute=True,
@@ -293,6 +293,12 @@ class SaleOrderLine(models.Model):
     qty_overage = fields.Float(compute="_compute_qty_overage", digits="Product Unit")
 
     # Analytic & Invoicing fields
+    invoice_policy = fields.Selection(
+        string="Invoicing Policy",
+        selection=[("order", "Ordered quantities"), ("delivery", "Delivered quantities")],
+        compute="_compute_invoice_policy",
+        search="_search_invoice_policy",
+    )
     qty_invoiced = fields.Float(
         string="Invoiced Quantity",
         compute="_compute_qty_invoiced",
@@ -394,6 +400,10 @@ class SaleOrderLine(models.Model):
         string="Collapse Composition", copy=True, default=False
     )  # Whether this section's lines will be hidden in reports and in the portal.
 
+    mandatory_product = fields.Boolean(
+        string="Is Product Mandatory", compute="_compute_mandatory_product"
+    )
+
     # === COMPUTE METHODS ===#
 
     @api.depends("order_partner_id", "order_id", "product_id")
@@ -587,10 +597,15 @@ class SaleOrderLine(models.Model):
             if line.display_type:
                 line.product_uom_qty = 0.0
 
-    @api.depends("product_id")
+    @api.depends("product_id", "display_type")
     def _compute_product_uom_id(self):
+        unit_uom = self.env["product.template"]._default_uom_id()
         for line in self:
-            if not line.product_uom_id or (line.product_id.uom_id.id != line.product_uom_id.id):
+            if line.display_type:
+                line.product_uom_id = False
+            elif not line.product_id and not line.is_downpayment:
+                line.product_uom_id = unit_uom
+            else:
                 line.product_uom_id = line.product_id.uom_id
 
     @api.depends("product_id.sale_line_warn_msg")
@@ -601,9 +616,7 @@ class SaleOrderLine(models.Model):
                 line.product_id.sale_line_warn_msg if has_warning_group else ""
             )
 
-    @api.depends(
-        "product_id", "product_id.uom_id", "product_id.uom_ids", "product_id.extra_uom_ids"
-    )
+    @api.depends("product_id")
     def _compute_allowed_uom_ids(self):
         for line in self:
             line.allowed_uom_ids = line.product_id._get_available_uoms()
@@ -622,8 +635,10 @@ class SaleOrderLine(models.Model):
                 taxes = None
                 if line.product_id:
                     taxes = line.product_id.taxes_id._filter_taxes_by_company()
-                if not line.product_id or not taxes:
-                    # Nothing to map
+                # if it's a productless line add default company's tax
+                if not taxes and not line.product_id and not line.display_type:
+                    taxes = company.account_sale_tax_id
+                if not taxes:
                     line.tax_ids = False
                     continue
                 fiscal_position = line.order_id.fiscal_position_id
@@ -1007,7 +1022,11 @@ class SaleOrderLine(models.Model):
             # For other delivery methods, they are expected to add their own quantities to the
             # quantities already provided by the `_prepare_qty_delivered` method, including
             # analytic lines quantities for reinvoiceable products.
-            if line.qty_delivered_method == "manual" and line.product_id.reinvoice_policy != "no":
+            if (
+                line.qty_delivered_method == "manual"
+                and line.product_id
+                and line.product_id.reinvoice_policy != "no"
+            ):
                 line.qty_delivered_method = "analytic"
 
     def _get_consu_qty_delivered_method(self):
@@ -1130,6 +1149,25 @@ class SaleOrderLine(models.Model):
             else:
                 line.qty_overage = 0
 
+    def _compute_invoice_policy(self):
+        for line in self:
+            if line.product_id:
+                line.invoice_policy = line.product_id.invoice_policy
+            else:
+                line.invoice_policy = line.company_id.sale_invoice_policy
+
+    def _search_invoice_policy(self, operator, value):  # noqa: PLR6301
+        product_sol_domain = Domain("product_id", "!=", False) & Domain(
+            "product_id.invoice_policy", operator, value
+        )
+        non_product_sol_domain = (
+            Domain("display_type", "=", False)
+            & Domain("product_id", "=", False)
+            & Domain("company_id.sale_invoice_policy", operator, value)
+        )
+
+        return product_sol_domain | non_product_sol_domain
+
     @api.depends("invoice_lines.move_id.state", "invoice_lines.quantity")
     def _compute_qty_invoiced(self):
         """Compute the invoiced quantity invoiced.
@@ -1221,7 +1259,7 @@ class SaleOrderLine(models.Model):
             if line.state == "sale" and not line.display_type:
                 if line.product_id.type == "combo":
                     combo_lines.add(line)
-                elif line.product_id.invoice_policy == "order" and not (
+                elif line.invoice_policy == "order" and not (
                     line.qty_overage and line.order_id.invoice_overages
                 ):
                     line.qty_to_invoice = line.product_uom_qty - line.qty_invoiced
@@ -1510,6 +1548,11 @@ class SaleOrderLine(models.Model):
                 elif line in sale_order_lines:
                     line.parent_id = last_sub or last_section
 
+    def _compute_mandatory_product(self):
+        self.mandatory_product = (
+            self.env["ir.config_parameter"].sudo().get_bool("sale.mandatory_product")
+        )
+
     # === CONSTRAINT METHODS ===#
 
     @api.constrains("combo_item_id")
@@ -1537,9 +1580,16 @@ class SaleOrderLine(models.Model):
     @api.onchange("product_id")
     def _onchange_product_id(self):
         if not self.product_id:
+            self.name = ""
             return
         self._reset_price_unit()
 
+    @api.onchange("product_template_id")
+    def _onchange_product_template_id(self):
+        if not self.product_template_id:
+            self.product_id = False
+            self.name = ""
+
     # === CRUD METHODS ===#
 
     @api.model_create_multi
@@ -1570,8 +1620,8 @@ class SaleOrderLine(models.Model):
                 line.product_id.with_context(
                     skip_qty_available_update=True
                 ).qty_available -= qty_delivered
-            if line.product_id and line.state == "sale":
-                msg = self.env._("Extra line with %s", line.product_id.display_name)
+            if not line.display_type and line.state == "sale":
+                msg = self.env._("Extra line with %s", line.product_id.display_name or line.name)
                 line.order_id.message_post(body=msg)
 
         return lines
@@ -2135,7 +2185,7 @@ class SaleOrderLine(models.Model):
         return (
             self.order_id._can_be_edited_on_portal()
             and not self.combo_item_id
-            and self.product_id != self.company_id.sale_discount_product_id
+            and (not self.product_id or self.product_id != self.company_id.sale_discount_product_id)
         )
 
     def _get_rounding(self):
